lib/srv/db/common/auth.go

```diff
@@ -0,0 +1,175 @@
+	"github.com/gravitational/teleport/lib/auth/native"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/service/rds/rdsutils"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+)
+
+// AuthConfig is the database access authenticator configuration.
+type AuthConfig struct {
+	// AuthClient is the cluster auth client.
+	AuthClient *auth.Client
+	// Credentials are the AWS credentials used to generate RDS auth tokens.
+	Credentials *credentials.Credentials
+	// RDSCACerts contains AWS RDS root certificates.
+	RDSCACerts map[string][]byte
+	// Clock is the clock implementation.
+	Clock clockwork.Clock
+	// Log is used for logging.
+	Log logrus.FieldLogger
+}
+
+// CheckAndSetDefaults validates the config and sets defaults.
+func (c *AuthConfig) CheckAndSetDefaults() error {
+	if c.AuthClient == nil {
+		return trace.BadParameter("missing AuthClient")
+	}
+	if c.Credentials == nil {
+		return trace.BadParameter("missing Credentials")
+	}
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+	if c.Log == nil {
+		c.Log = logrus.WithField(trace.Component, "db:auth")
+	}
+	return nil
+}
+
+// Auth provides utilities for creating TLS configurations and
+// generating auth tokens when connecting to databases.
+type Auth struct {
+	cfg AuthConfig
+}
+
+// NewAuth returns a new instance of database access authenticator.
+func NewAuth(config AuthConfig) (*Auth, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &Auth{
+		cfg: config,
+	}, nil
+}
+
+// GetRDSAuthToken returns authorization token that will be used as a password
+// when connecting to RDS and Aurora databases.
+func (a *Auth) GetRDSAuthToken(sessionCtx *Session) (string, error) {
+	a.cfg.Log.Debugf("Generating auth token for %s.", sessionCtx)
+	return rdsutils.BuildAuthToken(
+		sessionCtx.Server.GetURI(),
+		sessionCtx.Server.GetRegion(),
+		sessionCtx.DatabaseUser,
+		a.cfg.Credentials)
+}
+
+// GetTLSConfig builds the client TLS configuration for the session.
+//
+// For RDS/Aurora, the config must contain RDS root certificate as a trusted
+// authority. For onprem we generate a client certificate signed by the host
+// CA used to authenticate.
+func (a *Auth) GetTLSConfig(ctx context.Context, sessionCtx *Session) (*tls.Config, error) {
+	addr, err := utils.ParseAddr(sessionCtx.Server.GetURI())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	tlsConfig := &tls.Config{
+		ServerName: addr.Host(),
+		RootCAs:    x509.NewCertPool(),
+	}
+	// Add CA certificate to the trusted pool if it's present, e.g. when
+	// connecting to RDS/Aurora which require AWS CA.
+	if len(sessionCtx.Server.GetCA()) != 0 {
+		if !tlsConfig.RootCAs.AppendCertsFromPEM(sessionCtx.Server.GetCA()) {
+			return nil, trace.BadParameter("failed to append CA certificate to the pool")
```

Comment:
nit: this might be a confusing cause for the failure. How about stating that the certificates are invalid instead?